flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/WithinCheckpointFileMergingSnapshotManager.java

```diff
@@ -0,0 +1,97 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.filemerging;
+
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.core.fs.FSDataOutputStream;
+import org.apache.flink.runtime.state.CheckpointedStateScope;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Executor;
+
+/** A {@link FileMergingSnapshotManager} that merging files within a checkpoint. */
+public class WithinCheckpointFileMergingSnapshotManager extends FileMergingSnapshotManagerBase {
+
+    /** A dummy subtask key to reuse files among subtasks for private states. */
+    private static final SubtaskKey dummySubtaskKey = new SubtaskKey("dummy", -1, -1);
+
+    /**
+     * OutputStreams to be reused when writing checkpoint files. For WITHIN_BOUNDARY mode, physical
+     * files are NOT shared among multiple checkpoints.
+     */
+    private final Map<Tuple3<Long, SubtaskKey, CheckpointedStateScope>, PhysicalFile> fileMap;
+
+    public WithinCheckpointFileMergingSnapshotManager(String id, Executor ioExecutor) {
+        // currently there is no file size limit For WITHIN_BOUNDARY mode
+        super(id, ioExecutor);
+        fileMap = new HashMap<>();
+    }
+
+    @Override
+    @Nonnull
+    protected PhysicalFile getOrCreatePhysicalFileForCheckpoint(
+            SubtaskKey subtaskKey, long checkpointId, CheckpointedStateScope scope)
+            throws IOException {
+        // TODO: FLINK-32076 will add a file pool for each subtask key.
+        Tuple3<Long, SubtaskKey, CheckpointedStateScope> fileKey =
+                Tuple3.of(
+                        checkpointId,
+                        scope == CheckpointedStateScope.SHARED ? subtaskKey : dummySubtaskKey,
+                        scope);
+        PhysicalFile file;
+        synchronized (fileMap) {
+            file = fileMap.remove(fileKey);
+            if (file == null) {
+                file = createPhysicalFile(subtaskKey, scope);
+            }
+        }
+        return file;
+    }
+
+    @Override
+    protected void tryReusingPhysicalFile(
+            SubtaskKey subtaskKey, long checkpointId, PhysicalFile physicalFile)
+            throws IOException {
+        // TODO: FLINK-32076 will add a file pool for reusing.
+        CheckpointedStateScope scope = physicalFile.getScope();
+        Tuple3<Long, SubtaskKey, CheckpointedStateScope> fileKey =
+                Tuple3.of(
+                        checkpointId,
+                        scope == CheckpointedStateScope.SHARED ? subtaskKey : dummySubtaskKey,
+                        scope);
+        PhysicalFile current;
+        synchronized (fileMap) {
+            current = fileMap.putIfAbsent(fileKey, physicalFile);
+        }
+        if (syncAfterClosingLogicalFile) {
+            FSDataOutputStream os = physicalFile.getOutputStream();
+            if (os != null) {
+                os.flush();
```

Comment:
I don't think we need invoke both `flush` and `sync` here.
For HDFS, `sync` means flush out the data in client's buffer and persist data to the disk device all the way. which `flush` only flush out the data in client's user buffer. So I think it's enough to invoke `sync` only here. IIUC, it is similar to POSIX defines.
Not sure if we have to do that for any other storages.

Reply:
According to the JavaDoc of `org.apache.flink.core.fs.FSDataOutputStream::sync()`, the implementation should ensure all the data (including those in cache and buffer) can be persisted in non-volatile storage. So I removed the call of `flush()`.